### PR TITLE
Crosscompile python aarch64 wheels with dockcross

### PR DESCRIPTION
**Pull request by jtattermusch**
Enables building of aarch64 wheels via crosscompilation.
Uses similar tricks (overriding plat-name and EXT_SUFFIX) as https://github.com/protocolbuffers/protobuf/pull/8280.

Some limitations:
- currently twine check and auditwheel repair is being skipped
- still need to doublecheck whether the wheels work well

The nice thing is that crosscompilation isn't slower than the regular build (unlike building under emulation).

FTR a very similar technique could be used to crosscompile the armv7 and armv6  (currently under "linux_extra" and being built under an emulator which takes ages).

### CODE REVIEW

**Comment by jtattermusch** [comment]
@lidizheng  @gnossen this is not ready for merge yet, but good enough for first review pass.

**Comment by jtattermusch** [comment]
@janaknat please review as well.

File: tools/run_tests/artifacts/build_artifact_python.sh [hunk 1]
```diff
@@ -136,12 +153,15 @@ then
...
-venv/bin/python -m pip install "twine<=2.0"
-venv/bin/python -m twine check dist/* tools/distrib/python/grpcio_tools/dist/*
-rm -rf venv/
+if [ "$GRPC_SKIP_TWINE_CHECK" == "" ]
+then
+  # Ensure the generated artifacts are valid.
+  "${PYTHON}" -m pip install virtualenv
+  "${PYTHON}" -m virtualenv venv || { "${PYTHON}" -m pip install virtualenv==16.7.9 && "${PYTHON}" -m virtualenv venv; }
+  venv/bin/python -m pip install "twine<=2.0"
+  venv/bin/python -m twine check dist/* tools/distrib/python/grpcio_tools/dist/*
```
Comment (lidizheng) [inline tools/run_tests/artifacts/build_artifact_python.sh:162]: Is there a reason for skipping twine checks? We are following manylinux2014, should be supported by PyPI official packages?

File: src/python/grpcio/commands.py [hunk 2]
```diff
@@ -211,6 +212,20 @@ class BuildExt(build_ext.build_ext):
     }
     LINK_OPTIONS = {}
 
+    def get_ext_filename(self, ext_name):
+      # since python3.5, python extensions' shared libraries use a suffix that corresponds to the value
+      # of sysconfig.get_config_var('EXT_SUFFIX') and contains info about the architecture the library targets.
+      # E.g. on x64 linux the suffix is ".cpython-XYZ-x86_64-linux-gnu.so"
+      # When crosscompiling python wheels, we need to be able to override this suffix
+      # so that the resulting file name matches the target architecture and we end up with a well-formed
+      # wheel.
```
Comment (lidizheng) [inline src/python/grpcio/commands.py:221]: Does setuptool try to find compilers based on the file name? Good finding.

File: src/python/grpcio/commands.py [hunk 3]
```diff
@@ -211,6 +212,20 @@ class BuildExt(build_ext.build_ext):
...
+      filename = build_ext.build_ext.get_ext_filename(self, ext_name)
+      orig_ext_suffix = sysconfig.get_config_var('EXT_SUFFIX')
+      new_ext_suffix = os.getenv('GRPC_PYTHON_OVERRIDE_EXT_SUFFIX')
+      if new_ext_suffix and filename.endswith(orig_ext_suffix):
+        filename = filename[:-len(orig_ext_suffix)] + new_ext_suffix
```
Comment (gnossen) [inline src/python/grpcio/commands.py:226]: Nit: `filename = filename.replace(orig_ext_suffix, new_ext_suffix, 1)` might be more readable?

File: tools/run_tests/artifacts/artifact_targets.py [hunk 4]
```diff
@@ -144,8 +144,16 @@ def build_jobspec(self):
             environ['PYTHON'] = '/opt/python/{}/bin/python'.format(
                 self.py_version)
             environ['PIP'] = '/opt/python/{}/bin/pip'.format(self.py_version)
-            environ['GRPC_BUILD_GRPCIO_TOOLS_DEPENDENTS'] = 'TRUE'
-            environ['GRPC_BUILD_MANYLINUX_WHEEL'] = 'TRUE'
+            if self.arch == 'aarch64':
+                environ['GRPC_SKIP_TWINE_CHECK'] = 'TRUE'
```
Comment (gnossen) [inline tools/run_tests/artifacts/artifact_targets.py:148]: Why is this necessary?

File: tools/run_tests/artifacts/build_artifact_python.sh (hunk 1 above)
Comment (jtattermusch) [reply tools/run_tests/artifacts/build_artifact_python.sh:162]: We need to skip auditwheel and twine check for now, since when crosscompiling we're running on x64 image, but we got aarch64 wheel.  To perform checks, you'd likely need to run the check under qemu (available in the docker image), but also on aarch64 version of python etc.
The plan is to first test the aarch64 wheels manually. It should be possible to add back twine and auditwheel checks, but more work is needed.

File: tools/run_tests/artifacts/artifact_targets.py (hunk 4 above)
Comment (jtattermusch) [reply tools/run_tests/artifacts/artifact_targets.py:148]: Responded in the other comment.

**Comment by jtattermusch** [comment]
Artifact build results (for aarch64)
- passing for python3.8 and 3.9
- failure for python3.7 (`tools/run_tests/artifacts/build_artifact_python.sh: line 26: /opt/python/cp37-cp37/bin/pip: No such file or directory`) https://source.cloud.google.com/results/invocations/1bcb1668-6aa9-4916-9748-f291ec40d08e/targets/github%2Fgrpc/tests;query=manylinux2014;passed=true;group=tasks;test=build_artifact.python_manylinux2014_aarch64_cp37-cp37;row=6 

built wheels can be downloaded here:
https://console.cloud.google.com/storage/browser/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/pull_request/linux/grpc_build_artifacts/22486/20210212-055319

**Comment by jtattermusch** [comment]
Manual run of auditwheel show on arm64 machine:
```
python -m auditwheel show grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl

grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl is consistent
with the following platform tag: "manylinux_2_24_aarch64".

The wheel references external versioned symbols in these system-
provided shared libraries: libpthread.so.0 with versions
{'GLIBC_2.17'}, libm.so.6 with versions {'GLIBC_2.17'}, libc.so.6 with
versions {'GLIBC_2.17'}, libstdc++.so.6 with versions
{'GLIBCXX_3.4.17', 'GLIBCXX_3.4.15', 'GLIBCXX_3.4.14',
'GLIBCXX_3.4.18', 'CXXABI_1.3.5', 'GLIBCXX_3.4.11', 'GLIBCXX_3.4.9',
'GLIBCXX_3.4.20', 'CXXABI_1.3', 'GLIBCXX_3.4'}

This constrains the platform tag to "manylinux_2_24_aarch64". In order
to achieve a more compatible tag, you would need to recompile a new
wheel from source on a system with earlier versions of these
libraries, such as a recent manylinux image.
```

So looks like some work is still needed.  Not sure if the problem could be the too-new version of gcc being used (we do use 4.9.4 instead of the default 4.8 because gRPC doesn't compile under gcc 4.8 - see the dockerfile)

**Comment by jtattermusch** [comment]
> Manual run of auditwheel show on arm64 machine:
> 
> ```
> python -m auditwheel show grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl
> 
> grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl is consistent
> with the following platform tag: "manylinux_2_24_aarch64".
> 
> The wheel references external versioned symbols in these system-
> provided shared libraries: libpthread.so.0 with versions
> {'GLIBC_2.17'}, libm.so.6 with versions {'GLIBC_2.17'}, libc.so.6 with
> versions {'GLIBC_2.17'}, libstdc++.so.6 with versions
> {'GLIBCXX_3.4.17', 'GLIBCXX_3.4.15', 'GLIBCXX_3.4.14',
> 'GLIBCXX_3.4.18', 'CXXABI_1.3.5', 'GLIBCXX_3.4.11', 'GLIBCXX_3.4.9',
> 'GLIBCXX_3.4.20', 'CXXABI_1.3', 'GLIBCXX_3.4'}
> 
> This constrains the platform tag to "manylinux_2_24_aarch64". In order
> to achieve a more compatible tag, you would need to recompile a new
> wheel from source on a system with earlier versions of these
> libraries, such as a recent manylinux image.
> ```
> 
> So looks like some work is still needed. Not sure if the problem could be the too-new version of gcc being used (we do use 4.9.4 instead of the default 4.8 because gRPC doesn't compile under gcc 4.8 - see the dockerfile)

An attempt to auditwheel repair:
```
python -m auditwheel repair grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl || echo FAILED
INFO:auditwheel.main_repair:Repairing grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl
usage: __main__.py [-h] [-V] [-v] command ...
__main__.py: error: cannot repair "grpcio-1.37.0.dev0-cp38-cp38-manylinux2014_aarch64.whl" to "manylinux2014_aarch64" ABI because of the presence of too-recent versioned symbols. You'll need to compile the wheel on an older toolchain.
```

File: src/python/grpcio/commands.py (hunk 2 above)
Comment (jtattermusch) [reply src/python/grpcio/commands.py:221]: So when the python extension attempts to be loaded, the filename suffix needs to match the configuration of the python you're running.  The full docs is here: https://www.python.org/dev/peps/pep-3149/

**Comment by jtattermusch** [comment]
> Artifact build results (for aarch64)
> 
> * passing for python3.8 and 3.9
> * failure for python3.7 (`tools/run_tests/artifacts/build_artifact_python.sh: line 26: /opt/python/cp37-cp37/bin/pip: No such file or directory`) https://source.cloud.google.com/results/invocations/1bcb1668-6aa9-4916-9748-f291ec40d08e/targets/github%2Fgrpc/tests;query=manylinux2014;passed=true;group=tasks;test=build_artifact.python_manylinux2014_aarch64_cp37-cp37;row=6
> 
> built wheels can be downloaded here:
> https://console.cloud.google.com/storage/browser/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/pull_request/linux/grpc_build_artifacts/22486/20210212-055319

The python3.7 failure seems to be caused by a typo that's now fixed.

File: tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+# The aarch64 wheels are being crosscompiled to allow running the build
+# on x64 machine. The dockcross/manylinux2014-aarch64 image is a x86_64
+# image with crosscompilation toolchain installed.
+# Use an older version of dockcross image that has gcc4.9.4 because it was built
+# before https://github.com/dockcross/dockcross/pull/449
+FROM dockcross/manylinux2014-aarch64:20200929-608e6ac
+
+# Update the package manager
+RUN yum update -y && yum install -y curl-devel expat-devel gettext-devel openssl-devel zlib-devel
```
Comment (janaknat) [inline tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile:23]: Are these installing the aarch64 dev libs? When I tried it, it pulled in the x86 version.

File: tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+###################################
+# Install Python build requirements
+RUN /opt/python/cp35-cp35m/bin/pip install --upgrade cython
+RUN /opt/python/cp36-cp36m/bin/pip install --upgrade cython
+RUN /opt/python/cp37-cp37m/bin/pip install --upgrade cython
+RUN /opt/python/cp38-cp38/bin/pip install --upgrade cython
+RUN /opt/python/cp39-cp39/bin/pip install --upgrade cython
```
Comment (janaknat) [inline tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile:31]: Same question as above.

File: tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile (hunk 5 above)
Comment (jtattermusch) [reply tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile:23]: nope, this is installing x86 libs. I guess the reason why this doesn't matter is that they are actually not strictly required for a successful build (I copied the line from x64 dockerfile and didn't investigate whether they are really needed).

File: tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile (hunk 6 above)
Comment (jtattermusch) [reply tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64/Dockerfile:31]: it's a x64 cython, but that doesn't really matter it seems. Even the python itself is a x64 python. That seems correct as we're crosscompiling and we need to run the build on a x64 machine (and cython is what invokes the build commands).

File: tools/run_tests/artifacts/build_artifact_python.sh [hunk 7]
```diff
@@ -57,6 +58,7 @@ ${SETARCH_CMD} "${PYTHON}" setup.py sdist
 
 # Wheel has a bug where directories don't get excluded.
 # https://bitbucket.org/pypa/wheel/issues/99/cannot-exclude-directory
+# shellcheck disable=SC2086
 ${SETARCH_CMD} "${PYTHON}" setup.py bdist_wheel $WHEEL_PLAT_NAME_FLAG
```
Comment (lidizheng) [inline tools/run_tests/artifacts/build_artifact_python.sh:62]: Double quote ([SC2086](https://github.com/koalaman/shellcheck/wiki/SC2086)) seems applicable to this shell command. Quoting variables should fix the shellcheck complain.

File: tools/run_tests/artifacts/build_artifact_python.sh [hunk 8]
```diff
@@ -33,14 +33,33 @@ export GRPC_PYTHON_BUILD_EXT_COMPILER_JOBS=${GRPC_PYTHON_BUILD_EXT_COMPILER_JOBS
...
+  # conveniently provides the value in the AUDITWHEEL_PLAT env
+  WHEEL_PLAT_NAME_FLAG="--plat-name=$AUDITWHEEL_PLAT"
+
+  # override the value of EXT_SUFFIX to make sure the crosscompiled .so files in the wheel have the correct filename suffix
+  GRPC_PYTHON_OVERRIDE_EXT_SUFFIX="$(${PYTHON} -c 'import sysconfig; print(sysconfig.get_config_var("EXT_SUFFIX").replace("-x86_64-linux-gnu.so", "-aarch64-linux-gnu.so"))')"
+  export GRPC_PYTHON_OVERRIDE_EXT_SUFFIX
+
+  # Set to empty string to disable the option (see https://github.com/grpc/grpc/issues/24498)
+  # TODO: enable ASM optimizations for crosscompiled wheels
+  export GRPC_BUILD_WITH_BORING_SSL_ASM=""
```
Comment (lidizheng) [inline tools/run_tests/artifacts/build_artifact_python.sh:51]: Maybe link to the PR changing the BoringSSL ASM enable flag? https://github.com/grpc/grpc/pull/25444

**Review by lidizheng — APPROVED** [review]

File: src/python/grpcio/commands.py (hunk 3 above)
Comment (jtattermusch) [reply src/python/grpcio/commands.py]: what I have is a bit more complicated, but it's more robust. I only want to replace a substring if it's a suffix. Theoretically, you can have the name of the extension that contains the same string? (but yeah, it's quite unlikely).
I might just use replace.

File: tools/run_tests/artifacts/build_artifact_python.sh (hunk 7 above)
Comment (jtattermusch) [reply tools/run_tests/artifacts/build_artifact_python.sh:62]: That wouldn't work. WHEEL_PLAT_NAME_FLAG can be empty and in that case `"$WHEEL_PLAT_NAME_FLAG"` becomes `""` and an empty arg will be forcibly passed to the command (which does break it).

File: src/python/grpcio/commands.py (hunk 3 above)
Comment (gnossen) [reply src/python/grpcio/commands.py]: That's true. Leaving it up to you.

**Comment by jtattermusch** [comment]
Looks like the only problematic symbol is 
`0000000000000000      DF *UND*	0000000000000000  GLIBCXX_3.4.20 _ZSt24__throw_out_of_range_fmtPKcz` (and it seems to be used by absl`

**Comment by jtattermusch** [comment]
So far I haven't figured out how to make sure that the GCC 4.9 crosscompiler in the dockcross image uses symbols that are compliant with manylinux2014 (as discussed in https://github.com/grpc/grpc/pull/25418#issuecomment-778569369), but I used a workaround instead:
- when crosscompiling linux aarch64 wheels, use `-static-libstdc++` after which no libstdc++ are referenced (because they are linked statically) and the resulting wheels are manylinux2014 compatible (confirmed with manual run auditwheel)
- this approach increases the size of the binary wheels slightly (increase of size approx. 23MB to ~25MB), but that seems reasonable since we currently don't have a better way of crosscompiling for aarch64. The plan is to remove the workaround in the future, but for now having a set of aarch64 wheels is the priority.

@lidizheng  @gnossen   I think this PR is now ready to merge. PTAL

**Review by gnossen — APPROVED** [review]
Nice investigation!

**Review by lidizheng — APPROVED** [review]
Thanks for digging this!

**Comment by jtattermusch** [comment]
known failures:
https://github.com/grpc/grpc/issues/25594
